node/src/components/consensus/era_supervisor.rs

```diff
@@ -193,16 +202,56 @@ where
             .iter_past(current_era, bonded_eras * 3)
             .collect();
 
-        info!(?era_ids, "collecting switch blocks");
-
-        let effects = effect_builder
-            .collect_key_blocks(era_ids)
-            .event(move |key_blocks| Event::InitializeEras {
-                key_blocks: key_blocks.expect("should have all the switch blocks in storage"),
-                validators,
-                timestamp,
-                genesis_start_time,
-            });
+        // Asynchronously collect the information needed to initialize all recent eras.
+        let effects = async move {
+            info!(?era_ids, "collecting key blocks");
+            let key_blocks = effect_builder
+                .collect_key_blocks(era_ids.iter().cloned())
+                .await
+                .expect("should have all the key blocks in storage");
+
+            if current_era > activation_era_id + bonded_eras * 2 {
+                // All eras can be initialized using the key blocks only.
+                (key_blocks, Default::default())
+            } else {
+                // We need the validator set for the activation era from some protocol state.
+                let state_root_hash = if activation_era_id == current_era {
+                    // The activation era is the current one, so the initial state root hash
+                    // contains its validator set.
+                    initial_state_root_hash
+                } else if activation_era_id.is_genesis() {
+                    // To initialize the first era, we can use the global state of the first block.
+                    *effect_builder
+                        .get_block_at_height_local(0)
+                        .await
+                        .expect("missing block in genesis era")
+                        .state_root_hash()
+                } else {
+                    // The parent of the first key block contains the activation era validator set.
+                    let next_key_block = &key_blocks[&(activation_era_id + 1)];
+                    *effect_builder
+                        .get_block_at_height_local(next_key_block.height() - 1)
+                        .await
+                        .expect("missing block in activation era")
+                        .state_root_hash()
```

Comment:
If we special case'd the Genesis then we can use `key_block[activation_era_id]` here, right?

Reply:
No, because that key block is from before the upgrade, and the upgrade could have changed the validator set. That's why we need to read it from the global state from after the upgrade.

Reply:
Right, I wanted to suggest adding 1 to the `next_key_block.height()`:
```rust
let next_key_block_height = &key_blocks[&activation_era_id].height() + 1;
*effect_builder.get_block_at_height_local(next_key_block_height)
```

Reply:
Anyway, obviously not a blocker.

Reply:
Right, but above @fizyk20 said he prefers the current method: https://github.com/CasperLabs/casper-node/pull/1023#discussion_r585594894
Because then I'd have to add the special case for genesis, because there is no key block for era 0:
```rust
let next_key_block_height = if activation_era_id.is_genesis() {
    0
} else {
   &key_blocks[&activation_era_id].height() + 1
};
```

I'm fine with either and I'm happy to change it, if we can only agree on _something_!

Reply:
We already have `.is_genesis()` special case in so many places that I really don't mind one more. But sure, go ahead and merge it.

Reply:
I'll make another PR with just this change, then we can discuss it separately there. :+1:

Reply:
https://github.com/CasperLabs/casper-node/pull/1053